horace_core/sqw/@Experiment/Experiment.m

```diff
@@ -94,13 +94,78 @@
             %              object.
             %
             % Each argument can be a single object or array of objects.
-            if nargin == 0
-                % add one null instrument if using the vanilla constructor
-                % to satisfy the requirements of subsequent initialisation
+            
+            obj = obj@serializable();
+            
+            % initialising the compressed component containers.
+            % these may be overwritten if they are passed in as arguments
+            % below
+            obj.instruments_ = unique_references_container('GLOBAL_NAME_INSTRUMENTS_CONTAINER','IX_inst');
+            obj.detector_arrays_ = unique_references_container('GLOBAL_NAME_DETECTORS_CONTAINER','IX_detector_array');
+            obj.samples_ = unique_references_container('GLOBAL_NAME_SAMPLES_CONTAINER','IX_samp');
+            % expdata is not compressed and has been initialised as an
+            % empty array
+            
+            if nargin == 0 || (nargin == 1 && isempty(varargin{1}))
+                % no components in initialising arguments
+                % so finish
                 return;
+                
+            elseif nargin == 4
+                % arguments are the component arrays, 
+                % initialise them
+                detector_arrays = varargin{1};
+                instruments     = varargin{2};
+                samples         = varargin{3};
+                expdata         = varargin{4};
+
+                % define expdata first, as the size of this determines the
+                % number of runs which obj will hold
+                if isa(expdata, 'IX_experiment')
+                    % add to default compressed container
+                    % this must be of the right size to define number of
+                    % runs for the experiment - the other fields can handle
+                    % adding duplicates once they know this numberval,
+                    obj.expdata_ = expdata;
+                elseif isempty(expdata)
+                    % do nothing, leave default array container empty
+                else
+                    error('HORACE:Experiment:invalid_argument', ...
+                          'input is not empty or IX_experiment');
+                end
+                                
+                obj = obj.add_input_with_checks(detector_arrays, 'IX_detector_array');
+                
+                obj = obj.add_input_with_checks(instruments,     'IX_inst');
+                
+                obj = obj.add_input_with_checks(samples,         'IX_samp');
+                
+                                
+            elseif nargin == 1 
```

Comment:
I prefer increasing numbers of `nargin` for ifs

Reply:
this is correct